connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -0,0 +1,277 @@
+        final int mod = cls.getModifiers();
+        return !Modifier.isAbstract(mod) && !Modifier.isInterface(mod);
+    }
+
+    private static List<Path> pluginDirs(Path topDir) throws IOException {
+        DirectoryStream.Filter<Path> dirFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path path) throws IOException {
+                return Files.isDirectory(path);
+            }
+        };
+
+        List<Path> dirs = new ArrayList<>();
+        // Non-recursive for now
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(topDir, dirFilter)) {
+            for (Path dir : listing) {
+                dirs.add(dir);
+            }
+        }
+        return dirs;
+    }
+
+    private static List<URL> jarPaths(Path pluginDir) throws IOException {
+        DirectoryStream.Filter<Path> jarFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path file) throws IOException {
+                return file.toString().toLowerCase(Locale.ROOT).endsWith(".jar");
+            }
+        };
+
+        List<URL> jars = new ArrayList<>();
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(pluginDir, jarFilter)) {
+            for (Path jar : listing) {
+                jars.add(jar.toUri().toURL());
+            }
+        }
+        return jars;
+    }
+
+    public void addAlias(PluginDesc<?> plugin, String alias) {
+        SortedMap<PluginDesc<?>, PluginClassLoader> inner = pluginLoaders.get(plugin.className());
+        if (inner != null) {
+            pluginLoaders.putIfAbsent(alias, inner);
+        }
+    }
+
+    private <T> void addPlugins(Collection<PluginDesc<T>> plugins, PluginClassLoader loader) {
+        for (PluginDesc<T> plugin : plugins) {
+            String pluginClassName = plugin.className();
+            SortedMap<PluginDesc<?>, PluginClassLoader> inner = pluginLoaders.get(pluginClassName);
+            if (inner == null) {
+                inner = new TreeMap<>();
+                pluginLoaders.put(pluginClassName, inner);
+            }
+            inner.put(plugin, loader);
+        }
+    }
+
+    public void initLoaders() {
+        for (String path : pluginPaths) {
+            try {
+                Path pluginPath = Paths.get(path).toAbsolutePath();
+                if (Files.isDirectory(pluginPath)) {
+                    for (Path dir : pluginDirs(pluginPath)) {
+                        log.info("Loading dir: {}", dir);
+                        URL[] jars = jarPaths(dir).toArray(new URL[0]);
+                        // log.info("Loading jars: " + Arrays.toString(jars));
+                        PluginClassLoader loader = newPluginClassLoader(jars);
+                        log.info("Using loader: {}", loader);
+                        PluginScanResult plugins = scanPluginPath(loader, jars);
+
+                        if (plugins.isEmpty()) {
+                            inactivePaths.put(dir, null);
```

Comment:
Now that's an interesting point. This is here because I'm aiming to add a file watcher and allow plugin injection during runtime without a restart. I was planning to issue a separate PR for that, but that was the motivation behind keeping track of both active or inactive dirs (probably this will change anyways when a file watch is implemented). I'll remove for now.